### PR TITLE
Data definitions for a game

### PR DESCRIPTION
**Pull request by tdaga**
Based on [this](https://app.diagrams.net/#G14VG4lp67KOCk3s1GJwyiEdcumoSl7b3b) diagram. Stubs are commented out to pass the linter.

### CODE REVIEW

File: services/roomService/src/lib/Maze.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  /**
+   * Updates the scores on the leaderboard
+   */ 
+  // updateLeaderboard(): void {}
+
+  /**
+   * Adds a game to the Maze
+   */
+  // addGame(game: Game): void {}
```
Comment (lindojong) [inline services/roomService/src/lib/Maze.ts:43]: if we have addGame here we also probably need removeGame

File: services/roomService/src/types/GamePlayer.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  private _inMaze: boolean;
+
+  /** Whether the GamePlayer can be invited to a Game * */
+  private _enableInvite: boolean;
+
+  /** Whether the GamePlayer has an invite pending * */
+  private _invitePending: boolean;
+
+  /** The Game that this Player is part of */
+  private _game?: Game;
```
Comment (lindojong) [inline services/roomService/src/types/GamePlayer.ts:22]: if a player is in a game then are they always in the maze?

File: services/roomService/src/types/GamePlayer.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import Player from './Player';
+import Game from './Game';
+
+/**
+ * Each Player who is in a Game is represented by a GamePlayer object
+ */
+export default class GamePlayer extends Player {
+
+  /** The start time of the Player * */
+  private _startTime?: Date;
```
Comment (lindojong) [inline services/roomService/src/types/GamePlayer.ts:10]: should we also have an optional `private _finishTime?: Date` stored so we can calculate and persist an exact duration of time spent in maze (the final score)?

File: services/roomService/src/types/GamePlayer.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import Player from './Player';
+import Game from './Game';
+
+/**
+ * Each Player who is in a Game is represented by a GamePlayer object
+ */
+export default class GamePlayer extends Player {
```
Comment (lindojong) [inline services/roomService/src/types/GamePlayer.ts:7]: Are we replacing a standard `Player` with `GamePlayer` everywhere? If a player wants to replay the game in a new instance, will we be creating a new `GamePlayer` instance, or should we be able to reset/wipe all values?

File: services/roomService/src/types/GamePlayer.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  /**
+   * Returns whether this player has completed the maze
+   */ 
+  // finish(): boolean {}
+
+  /**
+   * Accepts an invite to a Game
+   */ 
+  // acceptInvite(inviter: Player): Game {}
+}
```
Comment (lindojong) [inline services/roomService/src/types/GamePlayer.ts:60]: i think i've realized why you wanted `hasWon` here lol sorry for being confusing, think it'd be good if you added it back in

**Review by lindojong — COMMENTED** [review]
I think this is a good start! I don't mind if it's merged in as-is since I think the foundation is good. I've left a bunch of questions for anyone on the team to help answer, but we can continue to add/remove methods as needs arise.

File: services/roomService/src/types/GamePlayer.ts (hunk 2 above)
Comment (kaseyl) [reply services/roomService/src/types/GamePlayer.ts:22]: That makes sense to me.  Are there any other instances where they'd be in a game but not in the maze?

File: services/roomService/src/types/GamePlayer.ts (hunk 3 above)
Comment (kaseyl) [reply services/roomService/src/types/GamePlayer.ts:10]: Not sure we'd need to store finish time as the calculated time duration could immediately be sent to the game/database after finishing.

**Review by lindojong — APPROVED** [review]

**Review by kaseyl — APPROVED** [review]

**Pull request merged**